packages/cardpay-sdk/contracts/addresses.ts

```diff
@@ -100,6 +100,7 @@ const GOERLI = {
   multiSendCallOnly: '0x40A2aCCbd92BCA938b02010E17A5b8929b49130D',
   moduleProxyFactory: '0x00000000000DC7F163742Eb4aBEf650037b1f588',
   metaGuard: '0xe2847462a574bfd43014d1c7BB6De5769C294691',
+  wrappedNativeToken: '0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6', // WETH
```

Comment:
I'm not sure if this is the correct place for storing this address, but we need it to look up the usd exchange rate on uniswap for the gas price in native token.

Reply:
so the relay server does the the price oracle conversion work internally in order to calculate the amount of token to reimburse the relay's EOA. it's a little more involved than just talking to the uniswap contract--there is a bit of math to do on top of that. We should take a look at how the relay server does this and do the javascript equivalent.

Reply:
@habdelra my rationale for using the WETH address is that I wanted to use the `exchangeRateOf(tokenAddress)` to get the native coin usd rate, but for that I needed the token address. But since native coin doesn't have an address I chose WETH as an alternative. But I'm really not sure if this makes sense. Will take a look at how the relayer is doing conversion, thanks!

Reply:
It's coming from this contract: https://github.com/cardstack/cardstack-module-scheduled-payment/blob/main/contracts/Exchange.sol

I saw @FadhlanR added it and I thought I could use it for this conversion

Reply:
Do you think we should port https://github.com/5afe/safe-relay-service/blob/master/safe_relay_service/tokens/models.py#L121?

Reply:
hmmm, i think there is more work to do then to make this work with a Uniswap contracts (perhaps that is what @FadhlanR is working on now). The `Exchange` contract had a bunch of cardpay protocol baggage though. it might be easier to just talk directly to uniswap contracts instead of having to go thru the `Exchange` contract.

Reply:
> Do you think we should port https://github.com/5afe/safe-relay-service/blob/master/safe_relay_service/tokens/models.py#L121?

No, i think _this_ is what you need to port for the uniswap oracle to work: https://github.com/safe-global/safe-eth-py/blob/master/gnosis/eth/oracles/oracles.py#L77-L467

Reply:
@habdelra From your explanation, I think you mentioned the card pay exchange contract. But what @jurgenwerk  referring to here is the exchange contract that I've made for the scheduled payment module and already used Uniswap V3 as source data to get the USD rate of the token.

I think the Idea in this PR is to use that Exchange contract to get the conversion rate between two tokens (in this case `native token to X token`). Because the `exchangeRateOf(tokenAddress)` function only provides conversion to USD, we need to have conversion rates from `native token to USD` and `X token to USD`. With that conversion rates, we can get a conversion rate from `native token to X token = native token to USD / X token to USD`. 

And we need the `wrappedNativeTokenAddress` here because the `exchangeRateOf` function receives the token address as a parameter and inside that function, there's a process to get the `uniswap pool address` based on the defined USD token address and token address.

Reply:
Agree with @FadhlanR that this were my thoughts behind this implementation. 

I'm wondering if we can use the uniswap sdk and fetch the price of gas token in weth like here: https://docs.uniswap.org/sdk/2.0.0/guides/pricing

This way we can avoid converting both tokens to usd rate first, and get a better handle on the token specifics such as number of decimals.

Reply:
thanks for the clarification @FadhlanR :+1: 

@jurgenwerk yeah, that's exactly what i was getting at--if we go to purely uniswap based token prices, is there even a reason to have any extra onchain logic like the exchange contract. and that's awesome that there is actually already an SDK for this so no need to reverse engineer what gnosis did.

Reply:
@habdelra uniswap has 2 SDK versions - one is v2 and the other v3 (https://github.com/Uniswap/v2-sdk/, https://github.com/Uniswap/v3-sdk/). @FadhlanR's Exhange contract mentioned above is using v3, but the v3 sdk says it's in alpha (check the v3 repo description) and that it might change significantly between patch versions. 

Do you think it's okay to use the v2 sdk for now, or should we keep versions consistent between the contract and the sdk? 

Just for demonstration purposes, here's how this looks using the v2 uniswap sdk:

```
export async function currentGasPriceV2(provider: JsonRpcProvider, gasTokenAddress: string) {
  let network = await provider.getNetwork();
  let gasToken = await Fetcher.fetchTokenData(network.chainId, gasTokenAddress, provider);
  let wrappedNativeToken = await Fetcher.fetchTokenData(
    network.chainId,
    getAddressByNetwork('wrappedNativeToken', network.name),
    provider
  );

  let pair = await Fetcher.fetchPairData(gasToken, wrappedNativeToken, provider);

  let route = new Route([pair], wrappedNativeToken);

  return route.midPrice.toSignificant(6);
}
```

Reply:
uniswap V2 is like the gold standard for liquidity pools. I suggest using that (that's also what the relay server uses if that helps)